lib/src/http.dart

```diff
@@ -343,58 +231,202 @@ Map parseJsonResponse(http.Response response) {
+        // skipping retries may lead to intermittent issues that could be fixed
+        // with a retry. Failing to retry intermittent issues is likely to cause
+        // customers to wrap pub in a retry loop which will not improve the
+        // end-user experience.
+
+        // TODO: should these return PubConnectException('msg', couldRetry: false)?
+        if (osError.isDnsError) {
+          fail('Could not resolve URL "${request.url.origin}".');
+        } else if (osError.isSslError) {
+          fail(
+              'Unable to validate SSL certificate for "${request.url.origin}".');
+        }
+      }
+
+      return e;
+    }, retryIf: (e, stackTrace) async {
+      // TODO: think about IOException. RetryClient used to catch IOException
+      // from _PubHttpClient.
+
+      if (e is PubConnectException && e.couldRetry) return true;
+      if (e is PubHttpException && e.couldRetry) return true;
+
+      return (e is PubHttpException && e.couldRetry) ||
+          e is HttpException ||
+          e is TimeoutException ||
+          e is FormatException ||
+          (retryIf != null && await retryIf(e, stackTrace));
+    }, onRetry: (exception, retryCount) async {
+      if (onRetry != null) {
+        await onRetry(exception, retryCount, request, managedResponse);
+      } else {
+        log.io('Retry #${retryCount + 1} for '
+            '${request.method} ${request.url}...');
+      }
+
+      if (retryCount != 3) return;
+      if (!_retriedHosts.add(request.url.host)) return;
+      log.message('It looks like ${request.url.host} is having some trouble.\n'
+          'Pub will wait for a while before trying to connect again.');
+    }, maxAttempts: maxAttempts);
+  }
+}
+
+extension on http.BaseResponse {
+  void throwIfError() {
+    // Retry if the response indicates a server error.
+    if ([500, 502, 503, 504].contains(statusCode)) {
+      throw PubHttpException(this as http.Response, couldRetry: true);
+    }
+
+    // 401 responses should be handled by the OAuth2 client. It's very
+    // unlikely that they'll be returned by non-OAuth2 requests. We also want
+    // to pass along 400 responses from the token endpoint.
+    var tokenRequest = request!.url == oauth2.tokenEndpoint;
+    if (statusCode < 400 ||
+        statusCode == 401 ||
+        (statusCode == 400 && tokenRequest)) {
+      return;
+    }
+
+    if (statusCode == 406 &&
+        request!.headers['Accept'] == pubApiHeaders['Accept']) {
+      fail('Pub ${sdk.version} is incompatible with the current version of '
+          '${request!.url.host}.\n'
+          'Upgrade pub to the latest version and try again.');
+    }
+
+    if (statusCode == 500 &&
+        (request!.url.host == 'pub.dartlang.org' ||
+            request!.url.host == 'storage.googleapis.com')) {
+      fail('HTTP error 500: Internal Server Error at ${request!.url}.\n'
+          'This is likely a transient error. Please try again later.');
+    }
```

Comment:
Do we need this?

Can't we just print a pretty message when mapping exceptions? Also do we really to special case this for pub.dev and gcs?

Reply:
Just saying, let's cleanup this and make it generic.

Reply:
Sounds good.